lib/segment/src/segment.rs

```diff
@@ -1220,6 +1264,83 @@ impl Drop for Segment {
     }
 }
 
+fn move_strip_prefix_all(dir: &Path, dest: &Path, prefix: &Path) -> OperationResult<()> {
+    let entries = dir
+        .read_dir()
+        .map_err(|err| failed_to_read_dir_error(dir, err))?;
+
+    for entry in entries {
+        let entry = entry.map_err(|err| failed_to_read_dir_error(dir, err))?;
+
+        let path = entry.path();
+
+        let name =
+            strip_prefix(&path, prefix).map_err(|err| failed_to_move_error(&path, dest, err))?;
+
+        let dest = dest.join(name);
+
+        if path.is_dir() && dest.exists() {
+            move_strip_prefix_all(&path, &dest, prefix)?;
+        } else {
+            if let Some(dir) = dest.parent() {
+                if !dir.exists() {
+                    fs::create_dir_all(dir).map_err(|err| {
+                        OperationError::service_error(format!(
+                            "failed to create {dir:?} directory: {err}"
+                        ))
+                    })?;
+                }
+            }
+
+            fs::rename(&path, &dest).map_err(|err| failed_to_move_error(name, &dest, err))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn failed_to_read_dir_error(dir: &Path, err: impl fmt::Display) -> OperationError {
+    OperationError::service_error(format!("failed to read {dir:?} directory: {err}"))
+}
+
+fn failed_to_move_error(path: &Path, dest: &Path, err: impl fmt::Display) -> OperationError {
+    OperationError::service_error(format!("failed to move {path:?} to {dest:?}: {err}"))
+}
+
+fn append_path_strip_prefix(
+    builder: &mut Builder<impl io::Write>,
+    path: &Path,
+    dest: &Path,
+    prefix: &Path,
+) -> OperationResult<()> {
+    let name = strip_prefix(path, prefix).map_err(|err| failed_to_append_error(path, err))?;
+    append_path(builder, path, &dest.join(name))
+}
+
+fn append_path(
+    builder: &mut Builder<impl io::Write>,
+    path: &Path,
+    dest: &Path,
+) -> OperationResult<()> {
+    builder
+        .append_path_with_name(path, dest)
+        .map_err(|err| failed_to_append_error(path, err))
+}
+
+fn failed_to_append_error(path: &Path, err: impl fmt::Display) -> OperationError {
+    OperationError::service_error(format!(
+        "failed to append {path:?} path to the segment snapshot archive: {err}"
+    ))
+}
+
+fn strip_prefix<'a>(path: &'a Path, prefix: &Path) -> OperationResult<&'a Path> {
+    path.strip_prefix(prefix).map_err(|err| {
+        OperationError::service_error(format!(
+            "failed to strip {prefix:?} prefix from {path:?}: {err}"
+        ))
+    })
+}
```

Comment:
Would it be possible to move this functions out of `segment.rs`? The file is quite big already

Reply:
Done.